### PR TITLE
Adding EDProducer for FastSim event weight (decay bug fix)

### PR DESCRIPTION
**Pull request by sbein**
Event weight needed to correct FastSim for bug with double decays performed for long-lived SM particles:

https://indico.cern.ch/event/1117915/contributions/4704537/attachments/2378157/4063536/FastSIm_LLSMParticleIssue3_red.pdf

### CODE REVIEW

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+FastSimWeightPR31285To36122::FastSimWeightPR31285To36122(const edm::ParameterSet& iConfig):
+  genCollection(iConfig.getParameter<edm::InputTag>("genCollection")),
+  genCollectionTok(consumes<edm::View<reco::GenParticle>>(genCollection)),
+  genJetTag(iConfig.getParameter<edm::InputTag>("genJetTag")),
+  genJetTok(consumes<edm::View<reco::GenJet>>(genJetTag)),
+  recJetTag(iConfig.getParameter<edm::InputTag>("recJetTag")),
+  recJetTok(consumes<edm::View<pat::Jet>>(recJetTag))  
+{
+    produces< double >("EventWeight");
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: If a producer only produces one thing, the name can be omitted for simplicity

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py [hunk 2]
```diff
@@ -203,6 +203,15 @@ def makeTreeFromMiniAOD(self,process):
...
         self.VarsDouble.append("genTops:weight(GenTopWeight)")
 
+        #Event weight to fix FastSim pre-UL bug
+        if self.fastsim:
+            process.FastSimWeightPR31285To36122 = cms.EDProducer("FastSimWeightPR31285To36122",
+                genCollection = cms.InputTag("prunedGenParticles"),
+                genJetTag = cms.InputTag('slimmedGenJets'),
+                recJetTag = cms.InputTag("slimmedJets")
+            )
+            self.VarsDouble.append("FastSimWeightPR31285To36122:EventWeight(FastSimWeightPR31285To36122)")
```
Comment (kpedro88) [inline TreeMaker/python/makeTreeFromMiniAOD_cff.py]: the branch name could be simplified to just `FastSimWeight`

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    edm::EDGetTokenT<edm::View<reco::GenJet>> genJetTok;    
+    edm::InputTag recJetTag;
+    edm::EDGetTokenT<edm::View<pat::Jet>> recJetTok;   
+    
+    
+   TH1F *hRatio_GenJetHadronPtGenJetHadronFlavorLt4;
+   TH1F *hRatio_GenJetHadronPtGenJetHadronFlavorEqEq4;
+   TH1F *hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5;
+   TAxis * xax;
+
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: remove unnecessary whitespace/linebreaks (throughout this file)

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+#include "DataFormats/PatCandidates/interface/Jet.h"
+#include "TLorentzVector.h"
+
+#include <vector>
+
+class FastSimWeightPR31285To36122 : public edm::global::EDProducer<> {
+
+public:
+    explicit FastSimWeightPR31285To36122(const edm::ParameterSet&);
+    ~FastSimWeightPR31285To36122() override;
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: unnecessary, remove

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(19,6.516129);
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(20,7.685393);
+   
+   
+}
+
+FastSimWeightPR31285To36122::~FastSimWeightPR31285To36122() {
+  // do anything here that needs to be done at destruction time
+  // (e.g. close files, deallocate resources etc.)
+}
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: unnecessary, remove

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    float originxy;   
+    float decayxy;       
+    int hadronFlavor;     
+    int pdgid;
+    
+
+    for(const auto& genJet : *genJets) 
+    {
+           if (!(genJet.pt()>30)) continue;
+           genJet_tlv.SetPxPyPzE(genJet.px(), genJet.py(), genJet.pz(), genJet.energy());
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: TLorentzVectors are unnecessary here. All CMSSW candidate object have `.p4()` accessors that return a more efficient 4-vector.

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+           jetHasOffendingGp_ = false;//must use
+           mindr = 99;
+           hadronFlavor = -11;
+           leadHadronPt = 1;
+           originxy = -1;
+           for(const auto& recJet : *recJets) 
+            {
+                recJet_tlv.SetPxPyPzE(recJet.px(), recJet.py(), recJet.pz(), recJet.energy());
+                dr = recJet_tlv.DeltaR(genJet_tlv);
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: Use https://github.com/cms-sw/cmssw/blob/CMSSW_10_2_X/DataFormats/Math/interface/deltaR.h with `.p4()` as noted above

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                    {
+                        if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
+                    else if (hadronFlavor==4 && pdgid>400 && pdgid<500)
+                    {
+                        if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
+                    else if (hadronFlavor==5 && pdgid>400 && pdgid<600)
+                    {
+                    if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: Use consistent indentation

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                if (!(dr<0.4 && dr<mindr)) continue;
+                mindr = dr;
+                hadronFlavor = std::max(1,recJet.hadronFlavour());
+                if (hadronFlavor==22) hadronFlavor = 1;
+                
+                for(const auto& iPart : *genPartCands) {
+                    if (!(iPart.pt()>15)) continue;
+                    pdgid = abs(iPart.pdgId());
+                    genPart_tlv.SetPxPyPzE(iPart.px(), iPart.py(), iPart.pz(), iPart.energy());
+                    if (!(genPart_tlv.DeltaR(genJet_tlv)<0.4)) continue;
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: as above, use `DataFormats/Math/interface/deltaR.h` with `.p4()`

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+                        if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
+                    else if (hadronFlavor==4 && pdgid>400 && pdgid<500)
+                    {
+                        if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
+                    else if (hadronFlavor==5 && pdgid>400 && pdgid<600)
+                    {
+                    if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: The logic of this block can be simplified to avoid repeating the same condition:
```cpp
if (iPart.pt()>leadHadronPt) {
    if ( (hadronFlavor<4 && pdgid>100 && pdgid<400) or
         (hadronFlavor==4 && pdgid>400 && pdgid<500) or
         (hadronFlavor==5 && pdgid>400 && pdgid<600) )
    {
        leadHadronPt = iPart.pt();
    }
}
```

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+                    else if (hadronFlavor==4 && pdgid>400 && pdgid<500)
+                    {
+                        if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }
+                    else if (hadronFlavor==5 && pdgid>400 && pdgid<600)
+                    {
+                    if(iPart.pt()>leadHadronPt) {leadHadronPt = iPart.pt();}
+                    }        
+                    if(!(iPart.numberOfDaughters()>1)) continue;
+                    originxy = sqrt(pow(iPart.vx(),2) + pow(iPart.vy(),2)); originxy = originxy;
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: delete `originxy = originxy;`

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                    if(!(originxy<2.16)) continue;
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: in principle, all numerical values (PDG IDs, DR/pt/other cuts, etc.) should be given as named constants for clarity and consistency

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+#include <DataFormats/HepMCCandidate/interface/GenParticle.h>
+#include "DataFormats/JetReco/interface/GenJet.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+#include "TLorentzVector.h"
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: This include can be eliminated if TLV usage below is removed (as noted in other comments)

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 2 above)
Comment (sbein) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: True, although the weight is only appropriate for productions created with releases created in the specified range of PR's. The weight *can* be used without creating a bias for other FastSim samples, but it will hurt the statistics unnecessarily.

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 1 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: removed argument

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 3 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: done

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 4 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: Are you referring to the 4 lines indicated? It seems this results in a compilation error, also when removing the lines in the following comment as well.

Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: (if I remove just line 30, then it's fine)

Comment (kpedro88) [reply Utils/src/FastSimWeightPR31285To36122.cc]: Then just set it inline: `~FastSimWeightPR31285To36122() override {}`

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 5 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: removed

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 10 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: thanks, that's better

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 11 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: thanks, it was a hack to overcome the unused variable error at some point.

File: Utils/src/FastSimWeightPR31285To36122.cc (hunk 13 above)
Comment (sbein) [reply Utils/src/FastSimWeightPR31285To36122.cc]: done

File: Utils/src/FastSimWeightPR31285To36122.cc [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+    float originxy;   
+    float decayxy;       
+    int hadronFlavor;     
+    int pdgid;
+    float genJetPtThreshold = 30;
+    float genParticlePtThreshold = 15;
+    float drThreshold = 0.4;
+    float justInsidePipe = 2.16;
+    float justOutsidePipe = 2.17;
+    float farOutsideCaloButNotTooFarAway = 2000;
```
Comment (kpedro88) [inline Utils/src/FastSimWeightPR31285To36122.cc]: these should all be `const`

**Comment by sbein** [comment]
ah shoot - can we undo this last commit? Some extra stuff inadvertently got added

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py [hunk 15]
```diff
@@ -3,23 +3,22 @@
...
+    
     # files to process
     import FWCore.PythonUtilities.LumiList as LumiList
-    from TreeMaker.TreeMaker.TMEras import TMeras
     process.maxEvents = cms.untracked.PSet(
         input = cms.untracked.int32(self.numevents)
     )
     process.source = cms.Source("PoolSource",
         fileNames = cms.untracked.vstring(self.readFiles),
+        secondaryFileNames=cms.untracked.vstring(self.readFiles_sidecar),
```
Comment (kpedro88) [inline TreeMaker/python/makeTreeFromMiniAOD_cff.py]: please completely remove the most recent commit 0958a9c5406a5e17f90c8a3f93793af0aadc5f44 which seems to merge unrelated changes

**Comment by kpedro88** [comment]
```
git reset --hard HEAD~1
git push -f origin Run2_2017_V17_FastSimWeightForBugFix
```
(assuming you used the name `origin` for the remote for `UhhCmsAnalysis`)

**Comment by kpedro88** [comment]
please squash these two commits together (the unrelated changes should not enter the official Run2_2017_V17 branch): 0958a9c5406a5e17f90c8a3f93793af0aadc5f44, 3a485411cf656f49131fdba84bacecc363c30a76

**Comment by sbein** [comment]
I deleted the fork and tried again. Sorry that was rather extreme - but I didn't want to do more damage.

**Comment by kpedro88** [comment]
superseded by #616
